Resolution for late-arriving hrrrdas file within the gsianl ex-script

EdwardColon-NOAA opened the pull request:
Added a conditional block to the that addresses late-arriving hrrrdas files in scripts/rtma3d/exrtma3d_gsianl.ksh.  If the current min past the hour exceeds the hrrrdas cutoff, the script reverts to the use of gdas ensemble files. If a catchup cycle is involved, the code defaults to the use of hrrrdas ensemble files. 

NOTE: Since the development machine has reverted back to the 35+ minute obs arrival time, the HRRRDAS_CUTOFF in rtma3d_pbspro_conus.xml should be set to at least 45-50 minutes only on parallels run on WCOSS2 development. On WCOSS2 production, 32 minutes should remain a viable cutoff.

EdwardColon-NOAA committed:
Resolved the late-arriving hrrrdas file logic within the gsianl ex-script and added stdout and OUTPUT file copy commands needed for archiving.

NOTE: Since the development machine has reverted back to the 35+ minute obs arrival time, the HRRRDAS_CUTOFF set in rtma3d_pbspro_conus.xml should be set to at least 40 minutes only on parallels set up on WCOSS development.

GangZhao-NOAA commented on scripts/rtma3d/exrtma3d_gsianl.ksh:
```diff
@@ -27,14 +27,15 @@ OBS_DIR=${DATAOBSHOME}
-#fi
+else
+HRRRDAS_STATE=0
+fi
```
@EdwardColon-NOAA @AnnetteGibbs-NOAA @ManuelPondeca-NOAA @MatthewMorris-NOAA 

hi Ed,

The comment line says "#reverting to gdas if current min exceeds cutoff min". So, for example, if it is 06:45, so CUMMIN=45.  Given HRRRDAS_CUTOFF=32 (set in rtma3d_pbspro_conus.xml), then 45 > 32,  "current min exceeds cutoff min", it goes to GDAS. 

Why does it go to GDAS, not to HRRRDAS,  if current min exceeds cutoff min? Is it because if current min exceeding cutoff min,  then HRRRDAS  ensemble is NOT available? But if exceeding, then current time should be a late time in that hour, then there should be more chance that hrrrdas ensemble are available, so it should go to HRRRDAS ensemble considering HRRRDAS is better than GDAS.     Or because it is late, then there is no enough time to running GSI analysis with HRRRDAS ensemble since it costs too much time, so it goes to GDAS to save computational time? I am confused. 

Thanks,

-Gang

AnnetteGibbs-NOAA replied:
@EdwardColon-NOAA Can you wait until Matt and I have had a chance to review your PR, if Gang approves, before you merge?  We should be able to review it on Monday.  Thanks!

EdwardColon-NOAA commented:
Hi Gang,

The way the  HRRRDAS_CUTOFF parameter works is that it sets a time limit to
receive all of the hrrrdas_ensemble files. The gsianl dependencies within
the conus xml script (rtma3d_pbspro_conus.xml) check that all 36 hrrrdas
ensemble files are present before processing. Occasionally there are
late-arriving hrrrdas files. If we were to wait for these, the gsi step
would be triggered late and the cycle would likely bleed into the next
hour. To avoid this, a time dependency is set in the xml file
(HRRRDAS_CUTOFF). On the production machine, that is set to 32 minutes so,
what that means is that the gsi step will begin to run at 32 minutes past
the hour even if all of the hrrrdas files are not present. Instead of using
the hrrrdas ensemble files (which are incomplete), the gsi will then use
the gdas ensemble files. This contingency will allow the cycle to complete
at a reasonable time without defaulting to 3dvar-only settings.

Since the development machine prepbufr/bufr obs are set back to default
arrival time setting in which they are received at 35+ minutes after the
hour, the HRRRDAS_CUTOFF variable has to be adjusted to take this extra
delay into account so, instead of 32 minutes past the hour, I suggest
closer to 45 - 50 minutes past the hour. Otherwise the processing is
exactly the same on each machine.

-Edward

On Fri, Apr 18, 2025 at 6:47 PM Gang Zhao ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In scripts/rtma3d/exrtma3d_gsianl.ksh
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#discussion_r2051214874>:
>
> >  HRRRDAS_STATE=${HRRRDAS_BEC}
> -#else
> -#HRRRDAS_STATE=0
> -#fi
> +else
> +HRRRDAS_STATE=0
> +fi
>
> @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA> @AnnetteGibbs-NOAA
> <https://github.com/AnnetteGibbs-NOAA> @ManuelPondeca-NOAA
> <https://github.com/ManuelPondeca-NOAA> @MatthewMorris-NOAA
> <https://github.com/MatthewMorris-NOAA>
>
> hi Ed,
>
> The comment line says "#reverting to gdas if current min exceeds cutoff
> min". So, for example, if it is 06:45, so CUMMIN=45. Given
> HRRRDAS_CUTOFF=32 (set in rtma3d_pbspro_conus.xml), then 45 > 32, "current
> min exceeds cutoff min", it goes to GDAS.
>
> Why does it go to GDAS, not to HRRRDAS, if current min exceeds cutoff min?
> Is it because if current min exceeding cutoff min, then HRRRDAS ensemble is
> NOT available? But if exceeding, then current time should be a late time in
> that hour, then there should be more chance that hrrrdas ensemble are
> available, so it should go to HRRRDAS ensemble considering HRRRDAS is
> better than GDAS. Or because it is late, then there is no enough time to
> running GSI analysis with HRRRDAS ensemble since it costs too much time, so
> it goes to GDAS to save computational time? I am confused.
>
> Thanks,
>
> -Gang
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#pullrequestreview-2779576541>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQETT7C3IF7JRY3K3BD22F6GBAVCNFSM6AAAAAB3NXEJKKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONZZGU3TMNJUGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Update exrtma3d_gsianl.ksh

EdwardColon-NOAA commented:
The following is a sample of how this scheme works:

/apps/ops/prod/nco/core/prod_util.v2.0.13/exec/mdate
 0.006 + cut -c11-12
 0.013 + CURRMIN=39.  <-------current minute
 0.013 + /apps/ops/prod/nco/core/prod_util.v2.0.13/exec/mdate
 0.013 + cut -c1-10
 0.018 + CURRTIME=2025041922 <-----current time string (YYYYMMDDHH)
 0.018 + STARTTIME=2025041922 <-------current cycle string (YYYYMMDDHH)
 0.018 + [ 2025041922 -gt 2025041922 ] <-------the current time string and
cycle string match so this isn't a catchup cycle
 0.018 + [ 39 -le 45 ] <------the current minute is compared to the hrrrdas
cutoff. It is less than the cutoff so the algorithm defaults to the use of
hrrrdas ensemble rather than gdas ensemble files
 0.018 + HRRRDAS_STATE=1 <----HRRRDAS state (1-hrrrdas ensemble, 0-gdas
ensemble)




On Sat, Apr 19, 2025 at 6:56 PM Edward Colon - NOAA Affiliate <
***@***.***> wrote:

> Hi Gang,
>
> The way the  HRRRDAS_CUTOFF parameter works is that it sets a time limit
> to receive all of the hrrrdas_ensemble files. The gsianl dependencies
> within the conus xml script (rtma3d_pbspro_conus.xml) check that all 36
> hrrrdas ensemble files are present before processing. Occasionally there
> are late-arriving hrrrdas files. If we were to wait for these, the gsi step
> would be triggered late and the cycle would likely bleed into the next
> hour. To avoid this, a time dependency is set in the xml file
> (HRRRDAS_CUTOFF). On the production machine, that is set to 32 minutes so,
> what that means is that the gsi step will begin to run at 32 minutes past
> the hour even if all of the hrrrdas files are not present. Instead of using
> the hrrrdas ensemble files (which are incomplete), the gsi will then use
> the gdas ensemble files. This contingency will allow the cycle to complete
> at a reasonable time without defaulting to 3dvar-only settings.
>
> Since the development machine prepbufr/bufr obs are set back to default
> arrival time setting in which they are received at 35+ minutes after the
> hour, the HRRRDAS_CUTOFF variable has to be adjusted to take this extra
> delay into account so, instead of 32 minutes past the hour, I suggest
> closer to 45 - 50 minutes past the hour. Otherwise the processing is
> exactly the same on each machine.
>
> -Edward
>
> On Fri, Apr 18, 2025 at 6:47 PM Gang Zhao ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In scripts/rtma3d/exrtma3d_gsianl.ksh
>> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#discussion_r2051214874>:
>>
>> >  HRRRDAS_STATE=${HRRRDAS_BEC}
>> -#else
>> -#HRRRDAS_STATE=0
>> -#fi
>> +else
>> +HRRRDAS_STATE=0
>> +fi
>>
>> @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA>
>> @AnnetteGibbs-NOAA <https://github.com/AnnetteGibbs-NOAA>
>> @ManuelPondeca-NOAA <https://github.com/ManuelPondeca-NOAA>
>> @MatthewMorris-NOAA <https://github.com/MatthewMorris-NOAA>
>>
>> hi Ed,
>>
>> The comment line says "#reverting to gdas if current min exceeds cutoff
>> min". So, for example, if it is 06:45, so CUMMIN=45. Given
>> HRRRDAS_CUTOFF=32 (set in rtma3d_pbspro_conus.xml), then 45 > 32, "current
>> min exceeds cutoff min", it goes to GDAS.
>>
>> Why does it go to GDAS, not to HRRRDAS, if current min exceeds cutoff
>> min? Is it because if current min exceeding cutoff min, then HRRRDAS
>> ensemble is NOT available? But if exceeding, then current time should be a
>> late time in that hour, then there should be more chance that hrrrdas
>> ensemble are available, so it should go to HRRRDAS ensemble considering
>> HRRRDAS is better than GDAS. Or because it is late, then there is no enough
>> time to running GSI analysis with HRRRDAS ensemble since it costs too much
>> time, so it goes to GDAS to save computational time? I am confused.
>>
>> Thanks,
>>
>> -Gang
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#pullrequestreview-2779576541>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACFXXQETT7C3IF7JRY3K3BD22F6GBAVCNFSM6AAAAAB3NXEJKKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONZZGU3TMNJUGE>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> -------------------------------------------------
> Edward Colón
> *Lynker* at NOAA/NWS/NCEP/EMC
> 5830 University Research Ct.
> Rm #2025
> Riverdale, MD 20737
> Office: (301) 683-3815
> Cell: (301) 213-3566
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

MatthewMorris-NOAA reviewed (commented):
Hi @EdwardColon-NOAA, I'm a little confused as to why this is needed. The XML file contains the dependency to automatically trigger the GSI step at 32 minutes if the 36 HRRRDAS members are missing, so why is additional logic needed in the ex-script? There was already existing logic to revert to GDAS files if there weren't enough HRRRDAS files available. However, from previous testing, this block of code didn't work. It is similar to what is used in the operational HRRR, however, so perhaps it is just incomplete or there is a bug there?  Should we instead follow the logic the HRRR uses?

Also, I don't think we need to wait until 45-50 minutes past the hour on the development machine. The late-arriving obs files are already accounted for in the prepobs (eventually GSI) depedency.

EdwardColon-NOAA commented:
I am running a test without the added coding block and a revised cutoff
time.

On Mon, Apr 21, 2025 at 10:19 AM MatthewMorris-NOAA <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Hi @EdwardColon-NOAA <https://github.com/EdwardColon-NOAA>, I'm a little
> confused as to why this is needed. The XML file contains the dependency to
> automatically trigger the GSI step at 32 minutes if the 36 HRRRDAS members
> are missing, so why is additional logic needed in the ex-script? There was
> already existing logic to revert to GDAS files if there weren't enough
> HRRRDAS files available. However, from previous testing, this block of code
> didn't work. It is similar to what is used in the operational HRRR,
> however, so perhaps it is just incomplete or there is a bug there? Should
> we instead follow the logic the HRRR uses?
>
> Also, I don't think we need to wait until 45-50 minutes past the hour on
> the development machine. The late-arriving obs files are already accounted
> for in the prepobs (eventually GSI) depedency.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#pullrequestreview-2781349407>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQCCPVXYZHQ2FDTOWCT22T45HAVCNFSM6AAAAAB3NXEJKKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOOBRGM2DSNBQG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566

EdwardColon-NOAA committed:
Update exrtma3d_gsianl.ksh

This script was adjusted to follow the logic posed in the operational HRRR gsianl exscript. The key change is that the number of gdas ensemble is always counted to provide a value for the variable "nummem" regardless if there are a sufficient number of hrrrdas files or not.  After testing over several cycles with and without 36 hrrrdas files, the now runs correctly in either case with it defaulting to gdas if all 36 hrrrdas ensemble file elements are not present.

MatthewMorris-NOAA commented on scripts/rtma3d/exrtma3d_gsianl.ksh:
```diff
@@ -157,7 +144,7 @@ else
 fi
 
-if [ "${envir}" = "lsf" ] || [ "${envir}" = "pbspro" ] && [ ${HRRRDAS_STATE} -eq 0 ] ; then #WCOSS
+#if [ "${envir}" = "lsf" ] || [ "${envir}" = "pbspro" ] && [ ${HRRRDAS_BEC} -eq 0 ] ; then #WCOSS
```
Why is this commented out?

MatthewMorris-NOAA reviewed (commented):
Looks good, just left a quick question. Also, have you run a test to confirm that the workflow will default to using GDAS data if the HRRRDAS isn't available?

EdwardColon-NOAA commented:
The conditional block needs to be commented out for two reasons:

1. The gdas ensemble file count needs to be made and stored in "nummem."
Otherwise, the gdas ensemble file processing will not trigger and will
default to 3dvar-only

2. In case the gdas ensemble is needed, the ensemble numbers will already
be ready for use. Otherwise, a special coding block would need to be
created for the instance where the hrrrdas ensemble file number count is
insufficient, HRRRDAS_BEC=1, and the GDAS ensemble file retrieval python
script would need to be invoked a second time in the script.

I tested the script and it functions as it should i.e. GDAS ensemble files
will be used if the HRRRDAS ensemble file count is incomplete. Otherwise,
if it is complete, the HRRRDAS ensemble files will be used in the analysis.
The HRRRDAS gsianl exscript takes the same approach.

-Edward

On Mon, Apr 21, 2025 at 5:25 PM MatthewMorris-NOAA ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Looks good, just left a quick question. Also, have you run a test to
> confirm that the workflow will default to using GDAS data if the HRRRDAS
> isn't available?
> ------------------------------
>
> In scripts/rtma3d/exrtma3d_gsianl.ksh
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#discussion_r2053014399>:
>
> > @@ -157,7 +144,7 @@ else
>    ${ECHO} "Warning: ${OBS_DIR}: satmar does not exist!"
>  fi
>
> -if [ "${envir}" = "lsf" ] || [ "${envir}" = "pbspro" ] && [ ${HRRRDAS_STATE} -eq 0 ] ; then #WCOSS
> +#if [ "${envir}" = "lsf" ] || [ "${envir}" = "pbspro" ] && [ ${HRRRDAS_BEC} -eq 0 ] ; then #WCOSS
>
> Why is this commented out?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NOAA_3drtma/pull/47#pullrequestreview-2782274872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFXXQF6QRINWLVE4UPSJRL22VO4PAVCNFSM6AAAAAB3NXEJKKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOOBSGI3TIOBXGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
-------------------------------------------------
Edward Colón
*Lynker* at NOAA/NWS/NCEP/EMC
5830 University Research Ct.
Rm #2025
Riverdale, MD 20737
Office: (301) 683-3815
Cell: (301) 213-3566
